samples/bluetooth/nrf_dm/src/service.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <bluetooth/services/ddfs.h>
+#include <random/rand32.h>
+#include <dm.h>
+
+#include "service.h"
+#include "peer.h"
+
+
+static int dm_ranging_mode_set(uint8_t mode)
+{
+	if (mode == BT_DDFS_DM_RANGING_MODE_RTT) {
+		peer_ranging_mode_set(DM_RANGING_MODE_RTT);
+	} else if (mode == BT_DDFS_DM_RANGING_MODE_MCPD) {
+		peer_ranging_mode_set(DM_RANGING_MODE_MCPD);
+	} else {
+		return -1;
+	}
+
+	printk("DM set ranging mode to %s\n", mode == BT_DDFS_DM_RANGING_MODE_RTT ? "RTT" : "MCPD");
+	return 0;
+}
+
+static int dm_config_read(struct bt_ddfs_dm_config *config)
+{
+	enum dm_ranging_mode ranging_mode;
+
+	if (!config) {
+		return -1;
+	}
+
+	ranging_mode = peer_ranging_mode_get();
+	if (ranging_mode == DM_RANGING_MODE_RTT) {
+		config->mode = BT_DDFS_DM_RANGING_MODE_RTT;
+	} else if (ranging_mode == DM_RANGING_MODE_MCPD) {
+		config->mode = BT_DDFS_DM_RANGING_MODE_MCPD;
+	} else {
+		return -1;
+	}
+
+	return 0;
+}
+
+static uint16_t meter_to_decimeter(float value)
+{
+	float val;
+
+	val = (value < 0 ? 0 : value) * 10;
+	return val < UINT16_MAX ? (uint16_t)val : UINT16_MAX;
+}
+
+void service_distance_measurement_update(const bt_addr_le_t *addr, const struct dm_result *result)
+{
+	struct bt_ddfs_distance_measurement measurement;
+	int err;
+
+	if (!addr || !result) {
+		return;
+	}
+
+	if (result->quality == DM_QUALITY_OK) {
+		measurement.quality = BT_DDFS_QUALITY_OK;
+	} else if (result->quality == DM_QUALITY_POOR) {
+		measurement.quality = BT_DDFS_QUALITY_POOR;
+	} else if (result->quality == DM_QUALITY_DO_NOT_USE) {
+		measurement.quality = BT_DDFS_QUALITY_DO_NOT_USE;
+	} else if (result->quality == DM_QUALITY_CRC_FAIL) {
+		measurement.quality = BT_DDFS_QUALITY_DO_NOT_USE;
+	} else {
+		measurement.quality = BT_DDFS_QUALITY_NONE;
+	}
+
+	if (result->ranging_mode == DM_RANGING_MODE_RTT) {
+		measurement.ranging_mode = BT_DDFS_DM_RANGING_MODE_RTT;
+		measurement.dist_estimates.rtt.rtt =
+				meter_to_decimeter(result->dist_estimates.rtt.rtt);
+	} else {
+		measurement.ranging_mode = BT_DDFS_DM_RANGING_MODE_MCPD;
+		measurement.dist_estimates.mcpd.ifft =
+				meter_to_decimeter(result->dist_estimates.mcpd.ifft);
+		measurement.dist_estimates.mcpd.phase_slope =
+				meter_to_decimeter(result->dist_estimates.mcpd.phase_slope);
+		measurement.dist_estimates.mcpd.rssi_openspace =
+				meter_to_decimeter(result->dist_estimates.mcpd.rssi_openspace);
+		measurement.dist_estimates.mcpd.best =
+				meter_to_decimeter(result->dist_estimates.mcpd.best);
+	}
+
+	bt_addr_le_copy(&measurement.bt_addr, addr);
+
+	err = bt_ddfs_distance_measurement_notify(NULL, &measurement);
+	if (err && err != -EACCES) {
+		printk("Failed to send distance measurement (err %d)\n", err);
+	}
+}
+
+void service_azimuth_elevation_simulation(void)
```

Comment:
I am wondering if this simulation is required. Can't the DF-related characteristic have just zero values? I know this is the common service for both direction finding and distance measurement, but this example effectively only performs distance measurements.